the-collab-lab/tcl-26-smart-shopping-list
# Add and read shopping list items

## Description

This PR adds a component AddItemForm to add shopping list items, and components ShoppingList and ShoppingListItem to display shopping list items. 

We also started to arrange things inside the ListView and AddItemView components to match the wireframe (including adding a barebones Header component), and added CSS classes for later to the files we were working on. 

### AddItemForm
We created a form with controlled components, which means the form fields update a state variable (formValues) every time the user changes anything. On form submission, the current values are formatted for the database, and defaults for the last purchase date and number of purchases are also set. The form relies on a listId value passed down from the App component, which is used to add the item to the correct list (the one corresponding to the user token. 

### ShoppingList and ShoppingListItem
Again relying on the listId value passed down from the App component, we use a useCollection Firebase hook to fetch the items from the correct list document in Firestore. Since it's going to be more complex later on, we separated out the list item into its own component to keep things neat and tidy.

### Changes to App component
In order to have a listId to pass to our new components and use when reading and writing to Firestore, we added a useEffect hook to App.js to, on initial mount, run a check for the localStorage token and fetch the corresponding List ID from Firestore.

## Related Issue

Closes #4

## Acceptance Criteria


## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before
![image](https://user-images.githubusercontent.com/67769490/126597607-8d44d610-d017-403d-a395-67895c2933e6.png)
![image](https://user-images.githubusercontent.com/67769490/126597667-b61dc6e6-1f47-4447-a545-fff136e4f212.png)

### After
![image](https://user-images.githubusercontent.com/67769490/126597354-391bd2e5-a67d-40e1-a5d2-6d8fed036594.png)
![image](https://user-images.githubusercontent.com/67769490/126597161-d54ce407-9794-40ce-93a8-2f26dbfe3651.png)


## Testing Steps / QA Criteria

1. Pull down this branch and check it out.
2. Run npm install to update dependencies.
3. Run npm start to start local server.
4. Add our test token (token: three word token) manually to localStorage to allow the app to connect with our test list in Firestore.
4. Fill out the "Add Item" form with a new item and submit. Check the list view and [the database](https://console.firebase.google.com/project/tcl-26-shopping-list/firestore/data/) to make sure it added.
